src/graphene_set.h

```diff
@@ -0,0 +1,336 @@
+        std::vector<uint64_t> orderedSetItems(itemRank.size(), 0);
+        for (size_t i = 0; i < itemRank.size(); i++)
+            orderedSetItems[itemRank[i]] = receiverSetItems[i];
+
+        return orderedSetItems;
+    }
+
+    static std::vector<unsigned char> EncodeRank(std::vector<uint64_t> items, uint16_t nBitsPerItem)
+    {
+        size_t nItems = items.size();
+        size_t nEncodedWords = int(ceil(nBitsPerItem * nItems / float(WORD_BITS)));
+        std::vector<unsigned char> encoded(nEncodedWords, 0);
+
+        // form boolean array (low-order first)
+        bool bits[nEncodedWords * WORD_BITS];
+        for (size_t i = 0; i < items.size(); i++)
+        {
+            uint64_t item = items[i];
+
+            assert(ceil(log2(item)) <= nBitsPerItem);
+
+            for (uint16_t j = 0; j < nBitsPerItem; j++)
+                bits[j + i * nBitsPerItem] = (item >> j) & 1;
+        }
+
+        // encode boolean array
+        for (size_t i = 0; i < nEncodedWords; i++)
+        {
+            encoded[i] = 0;
+            for (size_t j = 0; j < WORD_BITS; j++)
+                encoded[i] |= bits[j + i * WORD_BITS] << j;
+        }
+
+        return encoded;
+    }
+
+    static std::vector<uint64_t> DecodeRank(std::vector<unsigned char> encoded, size_t nItems, uint16_t nBitsPerItem)
+    {
+        size_t nEncodedWords = int(ceil(nBitsPerItem * nItems / float(WORD_BITS)));
+
+        // decode into boolean array (low-order first)
+        bool bits[nEncodedWords * WORD_BITS];
+        for (size_t i = 0; i < nEncodedWords; i++)
+        {
+            unsigned char word = encoded[i];
+
+            for (size_t j = 0; j < WORD_BITS; j++)
+                bits[j + i * WORD_BITS] = (word >> j) & 1;
+        }
+
+        // convert boolean to item array
+        std::vector<uint64_t> items(nItems, 0);
+        for (size_t i = 0; i < nItems; i++)
+        {
+            for (size_t j = 0; j < nBitsPerItem; j++)
+                items[i] |= bits[j + i * nBitsPerItem] << j;
+        }
+
+        return items;
+    }
+
+    uint64_t GetFilterSerializationSize() { return ::GetSerializeSize(*pSetFilter, SER_NETWORK, PROTOCOL_VERSION); }
+    uint64_t GetIbltSerializationSize() { return ::GetSerializeSize(*pSetIblt, SER_NETWORK, PROTOCOL_VERSION); }
+    uint64_t GetRankSerializationSize() { return ::GetSerializeSize(encodedRank, SER_NETWORK, PROTOCOL_VERSION); }
+    ~CGrapheneSet()
+    {
+        if (pSetFilter)
+        {
+            delete pSetFilter;
+            pSetFilter = nullptr;
+        }
+
+        if (pSetIblt)
+        {
+            delete pSetIblt;
+            pSetIblt = nullptr;
+        }
+    }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream &s, Operation ser_action)
+    {
+        READWRITE(ordered);
+        READWRITE(nReceiverUniverseItems);
```

Comment:
throw right here if nReceiverUniverseItems > LARGE_MEM_POOL_SIZE to limit attack surface analysis.